File: framework/languages/dotnet/samples/TicTacToe/Server/Play/Infrastructure/ZLink/Sessions/Handlers/AuthenticatePlaySessionHandler.cs
using TicTacToe.Server.Configuration;
using TicTacToe.Shared.Contracts;
using Zlink.Framework.Contracts.Actors;
using Zlink.Framework.Contracts.Channels;
using Zlink.Framework.Contracts.Streams;

namespace TicTacToe.Server.Play.Infrastructure.ZLink.Sessions.Handlers;

// --8<-- [start:doc-session-auth]
internal sealed class AuthenticatePlaySessionHandler(
    IZLinkActorManager actors,
    IZLinkRouteClient channels,
    ILogger<AuthenticatePlaySessionHandler> logger)
    : IZLinkSessionPacketHandler<IZLinkSessionContext, AuthenticateReq>
{
    public async ValueTask HandleAsync(
        IZLinkSessionContext context,
        ZLinkSessionDispatchContext dispatch,
        AuthenticateReq authenticate,
        CancellationToken cancellationToken)
    {
        logger.LogInformation(
            "play stream: authenticate requested. sessionId={SessionId}",
            context.SessionId);

        var accessToken = authenticate.AccessToken.Trim();
        if (string.IsNullOrWhiteSpace(accessToken))
            throw new InvalidOperationException("Authentication token is empty.");

        var authenticated = await channels.RequestToChannel(
                SampleChannels.Api,
                new AuthenticatePlayerReq(accessToken))
            .Async<AuthenticatePlayerRes>(cancellationToken);

        logger.LogInformation(
            "play stream: authenticate accepted. sessionId={SessionId}, player={ActorId}",
            context.SessionId,
            authenticated.Player.ActorId);

        await EnsureActorBoundAsync(
            context,
            authenticated.Player,
            cancellationToken);

        await context.Client.Reply(new AuthenticateRes(authenticated.Player))
            .Async(cancellationToken);
    }

    private async ValueTask EnsureActorBoundAsync(
        IZLinkSessionContext context,
        PlayerInfo player,
        CancellationToken cancellationToken)
    {
        logger.LogInformation(
            "play stream: creating actor before dispatch. sessionId={SessionId}, actor={ActorId}",
            context.SessionId,
            player.ActorId);
        var playerActor = (await actors.GetOrCreate(player.ActorId, SampleTypes.PlayerActor)
            .Request(player).Async(cancellationToken)) switch
        {
            ZLinkActorCreateResult.Existing value => value.Actor,
            ZLinkActorCreateResult.Created value => value.Actor,
            _ => throw new InvalidOperationException("Player Actor creation was rejected.")
        };
        logger.LogInformation(
            "play stream: binding actor to session. sessionId={SessionId}, actor={ActorId}",
            context.SessionId,
            player.ActorId);
        var boundActor = await context.Actors.BindOrGetAsync(
            playerActor,
            cancellationToken);

        logger.LogInformation(
            "play stream: actor bound to session. sessionId={SessionId}, actor={ActorId}",
            context.SessionId,
            boundActor.ActorId);
    }
}
// --8<-- [end:doc-session-auth]

File: framework/languages/node/samples/TicTacToe.Ts/Server/Play/Infrastructure/ZLink/Sessions/Handlers/authenticate-play-session-handler.ts
import { Inject, Injectable } from '@nestjs/common';
import { ZLINK_ACTOR_MANAGER, ZLINK_ROUTE_CLIENT } from '@zlink-systems/nestjs';
import { SampleNames } from '../../../../../Configuration/sample-settings';
import {
  PacketNames,
  authenticatePlayerReq,
  authenticateRes
} from '../../../../../../Shared/Contracts/messages';
import {
  ZLinkPacket,
  type ZLinkActorManager,
  type ZLinkRouteClient,
  type ZLinkMessage,
  type ZLinkSessionContext,
  type ZLinkSessionDispatchContext
} from '@zlink-systems/framework';
import type { AuthenticatePlayerRes, AuthenticateReq } from '../../../../../../Shared/Contracts/messages';

@Injectable()
@ZLinkPacket(PacketNames.authenticateReq)
// --8<-- [start:doc-session-auth]
class AuthenticatePlaySessionHandler {
  constructor(
    @Inject(ZLINK_ACTOR_MANAGER) private readonly actors: ZLinkActorManager,
    @Inject(ZLINK_ROUTE_CLIENT) private readonly api: ZLinkRouteClient
  ) {}

  async handle(context: ZLinkSessionContext, _dispatch: ZLinkSessionDispatchContext, payload: ZLinkMessage): Promise<void> {
    const request = payload.decode<AuthenticateReq>(Object as never);
    const authenticated = await this.api
      .requestToChannel(
        SampleNames.apiChannel,
        authenticatePlayerReq(request.accessToken)
      )
      .submit<AuthenticatePlayerRes>();
    const created = await this.actors
      .getOrCreate(authenticated.player.actorId, SampleNames.playerActorType)
      .inMesh(SampleNames.playSpotNode)
      .request(authenticated.player)
      .submit();
    if (created.status === 'rejected') {
      throw new Error(`Player actor '${authenticated.player.actorId}' creation was rejected.`);
    }
    const actorRef = created.actor;
    await context.actors.bindOrGet(actorRef);
    context.client.reply(authenticateRes(authenticated.player)).submit();
  }
}
// --8<-- [end:doc-session-auth]

export { AuthenticatePlaySessionHandler };
